mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -3129,3 +3132,221 @@ _PUBLIC_ struct emsmdbp_object *emsmdbp_object_ftcontext_init(TALLOC_CTX *mem_ct
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	bin = talloc_zero(mem_ctx, struct Binary_r);
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	bin->lpb = (uint8_t *) talloc_strdup(mem_ctx, xml);
+	bin->cb = strlen(xml);
+
+	*attach_bin = bin;
+
+	talloc_free(local_mem_ctx);
+
+	return MAPI_E_SUCCESS;
+}
+
+/**
+   \details Create a sharing metadata XML file for folder sharing invitation, request and response
+            and store it as attachment in the sharing object.
+            See [MS-OXSHARE] for this kind of messages and [MS-OXSHRMSG] for the XML format.
+
+   \param emsmdbp_ctx pointer to the emsmdb provider context
+   \param sharing_object the original sharing message object to dump the XML data from
+
+   \return MAPI_E_SUCCESS on success, a MAPI error code otherwise
+ */
+_PUBLIC_ enum MAPISTATUS emsmdbp_object_attach_sharing_metadata_XML_file(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *sharing_object)
+{
+	struct Binary_r		*attach_bin;
+	struct emsmdbp_object	*attach;
+	enum MAPISTATUS		*retvals = NULL;
+	enum MAPISTATUS		retval;
+	enum mapistore_error	ret;
+	struct SPropTagArray	*props;
+	struct SRow		aRow;
+	TALLOC_CTX		*mem_ctx;
+	uint32_t		contextID, aid;
+	void			**data_pointers;
+
+	/* Sanity checks */
+	OPENCHANGE_RETVAL_IF(!emsmdbp_ctx, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!sharing_object, MAPI_E_INVALID_PARAMETER, NULL);
+
+	if (!emsmdbp_is_mapistore(sharing_object)) {
+		return MAPI_E_NO_SUPPORT;
+	}
+
+	mem_ctx = talloc_named(NULL, 0, "attach_sharing_metadata");
+	OPENCHANGE_RETVAL_IF(!mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	/* 1. Check if the message is a IPM.Sharing message */
+	props = talloc_zero(mem_ctx, struct SPropTagArray);
+	OPENCHANGE_RETVAL_IF(!props, MAPI_E_NOT_ENOUGH_MEMORY, mem_ctx);
+	props->cValues = 1;
+	props->aulPropTag = talloc_zero(props, enum MAPITAGS);
+	OPENCHANGE_RETVAL_IF(!props->aulPropTag, MAPI_E_NOT_ENOUGH_MEMORY, mem_ctx);
+	props->aulPropTag[0] = PidTagMessageClass;
+
+	data_pointers = emsmdbp_object_get_properties(mem_ctx, emsmdbp_ctx, sharing_object, props, &retvals);
+	OPENCHANGE_RETVAL_IF(!data_pointers || retvals[0] != MAPI_E_SUCCESS, MAPI_E_NOT_FOUND, mem_ctx);
+
+	if (strncmp((char *)data_pointers[0], "IPM.Sharing", strlen("IPM.Sharing")) != 0) {
+		talloc_free(mem_ctx);
+		return MAPI_E_NOT_FOUND;
+	}
+
+	/* 2. Create the attachment */
+	attach = emsmdbp_object_attachment_init(mem_ctx, emsmdbp_ctx, sharing_object->object.message->messageID, sharing_object);
+	OPENCHANGE_RETVAL_IF(!attach, MAPI_E_NOT_ENOUGH_MEMORY, mem_ctx);
+
+	contextID = emsmdbp_get_contextID(sharing_object);
+
+	ret = mapistore_message_create_attachment(emsmdbp_ctx->mstore_ctx, contextID, sharing_object->backend_object, attach, &attach->backend_object, &aid);
+	OPENCHANGE_RETVAL_IF((ret != MAPISTORE_SUCCESS), mapistore_error_to_mapi(ret), mem_ctx);
+
+	/* 3. Set two properties to the attachment */
```

Comment:
THAT'S A LIE

three properties are set to the attachment, check line 3336 :trollface: